scripts/2024-10-01_import-new-geometries/main.py

```diff
@@ -0,0 +1,68 @@
+
+# The following parameters need to be adjusted to the correct values:
+filedir = ''
+filename = 'geometries.gpkg'
+output_filename = 'import_geometries.sql'
+output_sgsids_filename = 'sgsids.txt'
+
+# The following sgsids are not in the database, they were manually translated to the probably correct sgsid:
+sgsid_translation_dict = {}
+
+# Only use entries with Bemerkung equal to one of the following:
+allowed_bemerkung = ['erraten', 'erraten_B', 'neu', 'neu (?)', 'unsich kein B']
+
+def read_gdf(file: str, layer: str, allowed_bemerkung: list, translation_dict: dict):
+    '''Read a geopandas dataframe from a file and layer, filter out entries with Bemerkung not in allowed_bemerkung and translate the sgsid according to translationDict.'''
+    gdf = gpd.read_file(file, layer=layer)
+    print(gdf['Bemerkung'].value_counts())
+    gdf = gdf[(gdf['Bemerkung'].isin(allowed_bemerkung))]
+    gdf.replace({'IDSGS_neu': translation_dict}, inplace=True)
+    return gdf
+
+def write_insert_statements(f: TextIO, gdf: GeoDataFrame, geom_type: str):
+    for i, row in gdf.iterrows():
+        f.write(
+            f"INSERT INTO study_{geom_type} (asset_id, geom_quality_item_code, geom) VALUES ((SELECT asset_id FROM asset WHERE sgs_id = {row['IDSGS_neu']}), 'revised', ST_GeomFromText('{row['geometry'].wkt}', 2056));\n")
+
+def main():
+    path = os.path.join(filedir, filename)
+    gdf_points = read_gdf(file=path, layer='backup_20240829__points', allowed_bemerkung=allowed_bemerkung, translation_dict=sgsid_translation_dict)
+    gdf_lines = read_gdf(file=path, layer='backup_20240829__lines', allowed_bemerkung=allowed_bemerkung, translation_dict=sgsid_translation_dict)
+    gdf_polygons = read_gdf(file=path, layer='backup_20240829__polygons', allowed_bemerkung=allowed_bemerkung, translation_dict=sgsid_translation_dict)
+    gdf_lines = gdf_lines.explode() # explode multi-linestrings to linestrings
+    gdf_polygons = gdf_polygons.explode() # explode multi-polygons to polygons
+    gdf_points = gdf_points[~gdf_points['geometry'].is_empty] # remove empty geometries
+
+    sgs_ids = np.unique(np.concatenate(
+        [gdf_lines['IDSGS_neu'].unique(),
+         gdf_points['IDSGS_neu'].unique(),
+         gdf_polygons['IDSGS_neu'].unique()]))
+
+    np.savetxt(os.path.join(filedir, output_sgsids_filename), sgs_ids.astype(int), fmt='%d', delimiter=',')
+
+    # for each sgsid write a sql delete statement to delete all entries with this sgsid. Commit the commands every 1000 rows.
+    with open(os.path.join(filedir, output_filename), 'w') as f:
+        for i, sgsid in enumerate(sgs_ids):
+            f.write(
+                f"DELETE FROM study_location WHERE asset_id = (SELECT asset_id FROM asset WHERE sgs_id = {sgsid});\n")
+            f.write(f"DELETE FROM study_trace WHERE asset_id = (SELECT asset_id FROM asset WHERE sgs_id = {sgsid});\n")
+            f.write(f"DELETE FROM study_area WHERE asset_id = (SELECT asset_id FROM asset WHERE sgs_id = {sgsid});\n")
```

Comment:
I noticed that we don't have a unique constraint on `asset.sgs_id`. This means it is theorectically possible to reuse the same `sgs_id` on multiple assets. This would lead to the comparison on the nested SELECT here to throw an error.

This is currently not an issue, as we do not have any duplicate `sgs_id` in our database. However, I think we should either prepare for that possibility here, or check if we can add a unique constraint to the `sgs_id`, so that no problems will arise in the future.

Reply:
Created #283